pkg/pool-utils/contracts/RestrictedBalancerPoolToken.sol

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+
+import "./BalancerPoolToken.sol";
+import "./BasePoolAuthorization.sol";
+
+/**
+ * @title Extension of BalancerPoolToken with holder allowlist
+ * @author Balancer Labs
+ * @dev
+ * - Restricts both senders and recipients of all transfers (including mints
+ *   and burns) to addresses on an allowlist
+ * - Includes permissioned functions for adding/removing members to/from the
+ *   allowlist
+ */
+abstract contract RestrictedBalancerPoolToken is BalancerPoolToken, BasePoolAuthorization {
+    mapping(address => bool) private _allowedAddresses;
+
+    event AddressAdded(address member);
+    event AddressRemoved(address member);
+
+    constructor(string memory tokenName, string memory tokenSymbol, address owner)
+        BalancerPoolToken(tokenName, tokenSymbol)
+        BasePoolAuthorization(owner)
+    {
+        // Add zero address to allowlist to avoid blocking mints and burns
+        _allowedAddresses[address(0)] = true;
+    }
+
+    // Overrides
+
+    /**
+     * @dev Override to enforce address allowlist.
+     */
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256
+    ) internal virtual override {
+        _require(isAllowedAddress(from), Errors.ERC20_TRANSFER_FROM_PROHIBITED_ADDRESS);
+        _require(isAllowedAddress(to), Errors.ERC20_TRANSFER_TO_PROHIBITED_ADDRESS);
+    }
+
+    /**
+     * @dev Override to enforce address allowlist
+     */
+    function _isOwnerOnlyAction(bytes32 actionId) internal view virtual override returns (bool) {
+        return
+            (actionId == getActionId(this.addAllowedAddress.selector)) ||
+            (actionId == getActionId(this.removeAllowedAddress.selector));
+    }
+
+    // Public functions
+
+    /**
+     * @dev Verifies that a given address is allowed to hold tokens.
+     */
+    function isAllowedAddress(address member) public view returns (bool) {
+        return _allowedAddresses[member];
```

Comment:
```suggestion
        return member == address(0) && _allowedAddresses[member];
```

`_burn` sends tokens to the zero address, so the current implementation prevents burns unless we add this. Similarly, `_mint` sends from the zero address, so if we want to restrict `from` we'd also need it. This additionally removes these 'unnecessary' storage loads. (see https://docs.openzeppelin.com/contracts/3.x/api/token/erc20#ERC20-_beforeTokenTransfer-address-address-uint256-)

Alternatively, we could just override `_transfer` and `_mint`.

Reply:
Actually, I'd rather leave `isAllowedAddress` untouched, and have the `if` on the `_beforeTokenTransfer` hook, where we have context on the semantics of `from` and `sender`.

Reply:
> Alternatively, we could just override _transfer and _mint.

Did you read my text in the body of the PR?

Reply:
Yup, I was referrencing that (I should've added a 'as you pointed out' at the end). The point I tried to get accross is that my suggestion is simply an alternative to the mint and transfer override - both are equally good solutions.

Reply:
Sorry, I misunderstood your suggestion. I believe I've now implemented it as described.